### ISSUE TITLE
Creating multiple association table rows when saving a row on one side of the relationship

### ISSUE THREAD

**Issue by UltraBob** [miguelgrinberg/flasky#162]
Hi Miguel,

The book has been great I've been following along with it as I create a separate app, and it has proved invaluable.  I've now gotten all the way to the API chapter, and still haven't come across an answer to this.  I hope you don't mind if I ask for help or a pointer to the best place to find the answer.

I have a table (companies) with a many to many relationship via an association table (company_technologies) with another table (technologies).  I have a form for creating Companies that asks the user to select one or more technologies.  This all works fine, but when I get the results back, and try to assign things in the way we've previously done in the book I get an `AttributeError: 'int' object has no attribute '_sa_instance_state'` error.

How can one use the results of a SelectMultipleField to correctly create the required rows in the association table?

Thanks again for all the materials on Flask you've produced!

Bob

P.S. Here is an abbreviated [look at my models](https://gist.github.com/UltraBob/c74f75c0e36427856ab78f530fc89d4c)

and a [traceback](https://gist.github.com/182aa8e2c8f56c1ccfd63b56342d660d)

**Comment by miguelgrinberg** [comment]
Try treating your relationships as lists. For example, to add a country, do `company.countries.append(country)`.

**Comment by UltraBob** [comment]
Oh Ok.  I'm still not getting that to work, and wonder if my problem is the way I've set up my relationship.  I tried:

```
company.countries.append(form.countries.data)
```

```company.countries.append(form.countries)```

```
for id in form.countries.data:
            company.countries.append(id)
```

all of those returned `AttributeError: 'int' object has no attribute '_sa_instance_state'` or `AttributeError: 'SelectMultipleField' object has no attribute '_sa_instance_state'` or `AttributeError: 'list' object has no attribute '_sa_instance_state'`

and by the way, sorry, I don't know why I asked about the technologies relationship and then shared errors about the countries relationship, but I have 4 relationships like that, all set up pretty much exactly the same way.

**Comment by miguelgrinberg** [comment]
The `form.countries.data` list probably contains IDs for your countries, or maybe names. It does not contain the actual country models. For each of those, you need to load the model from the database, and then put the model in the `append()` call.

**Comment by UltraBob** [comment]
OK, I guess I was thinking too close to the database.  Since the country already exists, and what needs written is a new row in the lookup table containing the new company id and the existing country id.  I thought the country id was probably what I wanted to provide there.  Now that I try appending the Models one by one:

```
for id in form.countries.data:
            company.countries.append(Country.query.get(id))
```

I've finally gotten rid of the _sa_instance_state error and now have a `KeyError: 'companies'` error instead.

**Comment by miguelgrinberg** [comment]
This new error you are getting seems unrelated. I think it is a good sign, it means you are getting an error further along. Without a stack trace I can't really comment on the new error, my guess is that it is just a bug in your code.

**Comment by UltraBob** [comment]
Here is the [traceback](https://gist.github.com/5f80bd111e43bdbd944fcecb030336e3) I'll keep trying.

**Comment by UltraBob** [comment]
Here is the [traceback](https://gist.github.com/5f80bd111e43bdbd944fcecb030336e3) I'll keep trying.

It does seem to give me the same error whether I do company.countries.update(form.company.data) or company.countries.update([[Country.query.get(id) for id in form.company.data]]) so the change to update from append or = is what got me to this point and since I just tried that on a whim, because of the key error, I wonder if that isn't incorrect.

**Comment by miguelgrinberg** [comment]
Why do you have double brackets in your list comprehension?

And in any case, while using list comprehensions is great once you get everything to work, I recommend that unitl you get to that point you switch to a regular loop that spans multiple lines, as that will allow you to debug this more easily.

Can really tell for sure, but this last error seems to be that your `company` model is not part of the current database session. So I would look at how that company model was loaded.

**Comment by UltraBob** [comment]
Good question about why there were double brackets!  I have no idea, but that is likely a cause of part of my pain *headdesk*  I think I have some re-reading to do, I don't think I have fully grasped the ORM parts of this.  I'll also go read the sql alchemy ORM tutorial.  

As for loading the company model I have:

```
if form.validate_on_submit():
        company = Company()
```

Anyway, I'm sorry to have taken up so much of your time, and I really appreciate the help you've provided.

**Comment by miguelgrinberg** [comment]
This shouldn't be necessary, but you can try creating the company with all its properties except these many-to-many relationships. Save it to the database, then add the countries and other relationships.

**Comment by UltraBob** [comment]
OK, just moved db.session.add(company) up before starting to add form data and am on to different errors.  

I'll stop bugging you about this, but if I could can I ask for an example line of how you would add the results of a SelectMultipleField containing integer country ids to a company.countries field that actually represents a relationship to one side of a lookup table with columns company.id and country.id?  Since I don't know if I'm doing that right, it is tough to figure out whether the issue is there or with something else.

**Comment by miguelgrinberg** [comment]
This one you showed above is correct:

    for id in form.countries.data:
        company.countries.append(Country.query.get(id))

**Comment by UltraBob** [comment]
OK, thank you!  That gets me to a rather mysterious `KeyError: 'companies'` that seems to be triggered at `company.countries.append(Country.query.get(id))` where we aren't dealing with any `companies`, but maybe that means I need to also pass information about the company being created here?

Anyway thanks so much for your time here!

**Comment by miguelgrinberg** [comment]
I need to see the stack trace of the KeyError.

**Comment by UltraBob** [comment]
Ah, I just added it to that previous comment

**Comment by miguelgrinberg** [comment]
No, that was a `DetachedInstanceError`.

**Comment by UltraBob** [comment]
I mean [this one](https://gist.github.com/2c111ac4e871b1b69b5d0ce58ed8e3b1)

**Comment by miguelgrinberg** [comment]
Hmm. It's a weird error. Are you sure your relationships and backrefs are all correct? Search for the word "companies" in your models file, and see if that helps you spot any mistakes.

**Comment by UltraBob** [comment]
If I understand the traceback properly, the error is actually coming when it tries to load the backref defined here in the Companies class:

```
countries = db.relationship('Company',
                                secondary=company_countries,
                                backref=db.backref('companies', lazy='dynamic'),
                                lazy='dynamic')
```

which refers to this lookup table:

```
company_countries = db.Table('company_countries',
                             db.Column('company_id', db.Integer, db.ForeignKey('companies.id')),
                             db.Column('country_id', db.Integer, db.ForeignKey('countries.id'))
                             )
```

I'll go back through the info on setting those up, I don't immediately see problems, but I have no confidence that I would.

**Comment by miguelgrinberg** [comment]
Shouldn't this:

    countries = db.relationship('Company', ...)

be this instead?

    countries = db.relationship('Country', ...)

**Comment by UltraBob** [comment]
I'm such an idiot, thank you!  That error went away, and the errors I'm getting now are about the next relationship form that I haven't yet adapted to append, so it looks like that was the major malfunction.  Thanks again Miguel, I really appreciate your work!

**Comment by UltraBob** [comment]
HI Miguel, I'm going through the deployment chapter now, and wondered if it would be helpful to note what has changed with the Heroku process?  I'm reading the safari books online version of the book which perhaps they can update between printings, but I'm guessing it wouldn't help, I just thought I'd offer in case it is something I can give back.

**Comment by miguelgrinberg** [comment]
@UltraBob Updates to the e-books can happen when they get closer to running out of stock and need to do a reprint, and they let me know when that time comes so that I can get any updates I want in. They let me make updates to the book at any time, but those updates are only released at those reprint times. I have some updates on the Heroku chapter already made, but I'll have to do another pass and make sure there haven't been even more changes when the time comes.

**Comment by UltraBob** [comment]
OK, cool.  I thought so.
